File: cjeongmin/길_찾기_게임.js
const Node = (v) => ({ v, left: null, right: null });

const compareNodeFn = ({ node: a }, { node: b }) => {
    if (a[1] === b[1]) return a[0] - b[0];
    return b[1] - a[1];
};

const insert = (root, nodeinfo) => {
    if (root === null) return Node(nodeinfo);

    if (nodeinfo.node[0] < root.v.node[0]) root.left = insert(root.left, nodeinfo);
    else if (nodeinfo.node[0] > root.v.node[0]) root.right = insert(root.right, nodeinfo);

    return root;
};

const preorder = (root) => {
    if (root === null) return [];
    return [root.v.index].concat(preorder(root.left), preorder(root.right));
};

const postorder = (root) => {
    if (root === null) return [];
    return postorder(root.left).concat(postorder(root.right), [root.v.index]);
};

function solution(nodeinfo) {
    const nodes = nodeinfo.map((node, index) => ({ node, index: index + 1 })).sort(compareNodeFn);

    const root = Node(nodes[0]);

    for (let i = 1; i < nodes.length; i++) insert(root, nodes[i]);

    return [preorder(root), postorder(root)];
}

File: jiho/길찾기게임.py
import sys
sys.setrecursionlimit(10000)

class Node:
    def __init__(self, x, y, i):
        self.x = x
        self.y = y
        self.i = i
        self.left = None
        self.right = None

def insert(root, node):
    if root == None:
        return node
    if node.x < root.x:
        root.left = insert(root.left, node)
    elif node.x > root.x:
        root.right = insert(root.right, node)
    return root
    
def preorder(node):
    if node == None:
        return []
    return [node.i] + preorder(node.left) + preorder(node.right)

def postorder(node):
    if node == None:
        return []
    return postorder(node.left) + postorder(node.right) + [node.i]
    
def solution(nodeinfo):
    nodes = [Node(x, y, i+1) for i, (x, y) in enumerate(nodeinfo)]
    nodes.sort(key=lambda node: (-node.y, node.x))
    tree = nodes[0]
    for node in nodes[1:]:
        insert(tree, node)
    return [preorder(tree), postorder(tree)]